### ISSUE TITLE
[Vector Distribution] Layout Choices

### ISSUE THREAD

**Issue by Groverkss** [iree-org/iree#16127]
This issue tracks the resources + discussion for deciding how the layout should look like for Vector Distribution.

**Comment by Groverkss** [comment]
Triton Layouts: https://github.com/openai/triton/blob/597ffcf6193c6f833df7fb10f3be25979ca22cda/include/triton/Dialect/TritonGPU/IR/TritonGPUAttrDefs.td

CuTe Layouts: https://github.com/NVIDIA/cutlass/blob/main/media/docs/cute/01_layout.md

Graphene IR Layouts: https://dl.acm.org/doi/pdf/10.1145/3582016.3582018

**Comment by kuhar** [comment]
Examples:

1. How can we express the layout for SIMT-level i8 matmul lowered to a sequence of 16 dot product operations: `vector<4xi8>, vector<4xi8> -> i32`?

```mlir
%11 = scf.for %arg4 = %c0 to %c512 step %c16 iter_args(%arg5 = %10) -> (tensor<4x4xi32>) {
    %extracted_slice_0 = tensor.extract_slice %6[%arg0, %arg4] [4, 16] [1, 1] : tensor<8x512xi8> to tensor<4x16xi8>
    %extracted_slice_1 = tensor.extract_slice %7[%arg4, %arg2] [16, 4] [1, 1] : tensor<512x32xi8> to tensor<16x4xi8>
    %12 = vector.transfer_read %extracted_slice_0[%c0, %c0], %c0_i8 {in_bounds = [true, true]} : tensor<4x16xi8>, vector<4x16xi8>
    %13 = vector.transfer_read %extracted_slice_1[%c0, %c0], %c0_i8 {in_bounds = [true, true]} : tensor<16x4xi8>, vector<16x4xi8>
    %14 = vector.transfer_read %arg5[%c0, %c0], %c0_i32 {in_bounds = [true, true]} : tensor<4x4xi32>, vector<4x4xi32>
    %15 = arith.extsi %12 : vector<4x16xi8> to vector<4x16xi32>
    %16 = arith.extsi %13 : vector<16x4xi8> to vector<16x4xi32>
    %17 = vector.contract
            {indexing_maps = [affine_map<(d0, d1, d2) -> (d0, d2)>,
                              affine_map<(d0, d1, d2) -> (d2, d1)>,
                              affine_map<(d0, d1, d2) -> (d0, d1)>],
             iterator_types = ["parallel", "parallel", "reduction"],
             kind = #vector.kind<add>
            }
           %15, %16, %14 : vector<4x16xi32>, vector<16x4xi32> into vector<4x4xi32>
    %18 = vector.transfer_write %17, %arg5[%c0, %c0] {in_bounds = [true, true]} : vector<4x4xi32>, tensor<4x4xi32>
    scf.yield %18 : tensor<4x4xi32>
}
```

Would it be:
```
    %15 = ... : vector<4x16xi32>  // <r=<[BATCHX, VECTORX], [4, 4]>, c=<[BATCHY], [4]>>
    %16 = ... : vector<16x4xi32>  // <r=<[BATCHY], [4]>, c=<[BATCHX, VECTORX], [4, 4]>>
    %17 = ... : vector<4x4xi32>  // <r=<[BATCHX], [4]>, c=<[BATCHY], [4]>>
```

![image](https://github.com/openxla/iree/assets/4612584/f0a58a3d-1e24-4477-b1d6-5ed36efee415)


2. SIMT distribution for matvec, assuming we want each subgroup to process 8xf16 elements per read?
```mlir
%workgroup_id_x = hal.interface.workgroup.id[0] : index
%3 = scf.for %arg0 = %c0 to %c4096 step %c512 iter_args(%arg1 = %cst) -> (vector<512xf16>) {
    %7 = vector.transfer_read %0[%c0, %arg0], %cst_1 {in_bounds = [true]} : memref<1x4096xf16, #hal.descriptor_type<storage_buffer>>, vector<512xf16>
    %8 = vector.transfer_read %1[%workgroup_id_x, %arg0], %cst_1 {in_bounds = [true]} : memref<32000x4096xf16, #hal.descriptor_type<storage_buffer>>, vector<512xf16>
    %9 = arith.mulf %7, %8 : vector<512xf16>
    %10 = arith.addf %arg1, %9 : vector<512xf16>
    scf.yield %10 : vector<512xf16>
}
%4 = vector.broadcast %3 : vector<512xf16> to vector<1x1x512xf16>
%5 = vector.multi_reduction <add>, %4, %cst_0 [2] : vector<1x1x512xf16> to vector<1x1xf16>
%6 = vector.extract %5[0] : vector<1xf16> from vector<1x1xf16>
vector.transfer_write %6, %2[%c0, %workgroup_id_x] {in_bounds = [true]} : vector<1xf16>, memref<1x32000xf16, #hal.descriptor_type<storage_buffer>>
```

Would it be:
```
%7 = ... : vector<512xf16>  // <[LANEX, VECTORX], [64,8]>
%8 = ... : vector<512xf16>  // <[LANEX, VECTORX], [64,8]> 
```

3. SIMT distribution for matvec, assuming we want each subgroup to process 4 rows and 8xf16 elements per row, and for the multi-reduction to reduce 2 elements at one?
```mlir
%workgroup_id_x = hal.interface.workgroup.id[0] : index
%3 = affine.apply affine_map<()[s0] -> (s0 * 4)>()[%workgroup_id_x]
%4 = scf.for %arg0 = %c0 to %c4096 step %c512 iter_args(%arg1 = %cst) -> (vector<4x512xf16>) {
    %8 = vector.transfer_read %0[%c0, %arg0], %cst_1 {in_bounds = [true, true], permutation_map = affine_map<(d0, d1) -> (0, d1)>} : memref<1x4096xf16, #hal.descriptor_type<storage_buffer>>, vector<4x512xf16>
    %9 = vector.transfer_read %1[%3, %arg0], %cst_1 {in_bounds = [true, true]} : memref<32000x4096xf16, #hal.descriptor_type<storage_buffer>>, vector<4x512xf16>
    %10 = arith.mulf %8, %9 : vector<4x512xf16>
    %11 = arith.addf %arg1, %10 : vector<4x512xf16>
    scf.yield %11 : vector<4x512xf16>
}
%5 = vector.broadcast %4 : vector<4x512xf16> to vector<1x4x512xf16>
%6 = vector.multi_reduction <add>, %5, %cst_0 [2] : vector<1x4x512xf16> to vector<1x4xf16>
```

Would it be:
```
%8 = ... : vector<4x512xf16>  // <r=<[LANEX, VECTORX], [64,8]>, c=<[BATCHY], [4]>>
%9 = ... : vector<4x512xf16>  // <r=<[LANEX, VECTORX], [64,8]>, c=<[BATCHY], [4]>>
...
%6 = ... : vector<1x4x512xf16>  // <r=<LANEX, VECTORX, [64, 8], c=<[BATCHY, VECTORY], [2, 2]>>
```

**Comment by qedawkins** [comment]
Some follow up comments on the discussion this morning, I'd like to start using this layout distribution for things other than matmul and that requires a bit of decoupling from the hard coded enums we currently have in-tree. The way that I am interpreting a vector "layout" is as a way to map a vector type from one virtual grid of indices to another. The set of indices in place today are
```
BATCHX, BATCHY, BATCHZ
LANEX, LANEY, LANEZ
VECTORX, VECTORY, VECTORZ
```
Alongside a list of sizes for each identifier per vector dimension. For distribution, the batch dimensions are used like outer loop dimensions (i.e. vector operations are unrolled along them), lane dimensions are distributed to virtual lane indices, and vector dimensions describe the vector type of the distributed shape. Distribution maps this to
```
for bz in [0, batch_dim_z] {
  for by in [0, batch_dim_y] {
    for bx in [0, batch_dim_x] {

      forall lz in [0, lane_dim_z] {
        forall ly in [0, lane_dim_y] {
          forall lz in [0, lane_dim_x] {

            vector<vector_dim_z x vector_dim_y x vector_dim_x>

          }
        }
      }

    }
  }
}
```

This has two immediate limitations. First, having only 3 lane(/batch/vector) indices limits the number of dimensions we can distribute to. Second, by only having lane indices means distribution to warps must happen separately. The disadvantage of this is for any given simple load, we have to first introduce warp ids and then rely on later cleanup patterns to not introduce extra arithmetic. For example

```mlir
%wx = gpu.warp_id x
vector.transfer_read %src[%wx] : vector<128xf32>
```
where after distribution w/ subgroup size 32, we get
```mlir
%tx = gpu.thread_id x
%lane_id = %tx % 32
%wx = gpu.warp_id x
%idx = %wx * 32 + %lane_id
vector.transfer_read %src[%idx] : vector<4xf32>
```
Where if we just distributed directly to threads, then we would trivially get `%idx = gpu.thread_id x`.

This gives two wants
1. Layout capable of representing both warp + lane and thread semantics
    a. Lane ids could just be treated as additional virtual thread indices if necessary, but I suspect the separation will be a nice convenience.
2. Arbitrary counts of distribution dimensions, i.e. not limited to X/Y/Z

The attribute I have in mind looks like the following
```
#layout = #iree_vector_ext.distributed<
  // A list of shapes representative of a shape_cast on the original vector shape.
  // This can probably be restricted to shape expansion only as an initial step.
  shape = [d0, d1, ..., dn],

  // Optional broadcasted shape
  bcast = [d0 * 2, d1 * 4, ..., dn]

  // An optional permutation of those dimensions. This is to achieve the x/y/z
  // ordering that the current layout does.
  perm  = [0, 3, 4, 1, 2, 5],

  // A list of dimension types, but this time just BATCH/LANE/blah, as the dimensions
  // here are implicitly ordered from outer most to inner most. Could do further
  // elision of outer most dims we want to just unconditionally distribute to batch
  // dimensions.
  types = [BATCH/LANE/THREAD/VECTOR labels],
>
```

We can derive the current layout from this one, and it keeps it easy to introduce
new dimension types. It also gives a good amount of opportunity for elision of
different parts of the layout for common cases.

Also as a note of personal experience, I found the per-dim layout approach a bit
unwieldy hence why I think flattening it could be nicer to work with. Also composes
better if we end up wanting to collapse unnecessary dimensionality as a part of the
layout (could help make some indexing easier).

For Jakub's matvec example this could be

```
%8 = ... : vector<4x512xf16>  // <[4, 64, 8], [BATCH, LANE, VECTOR]>
%9 = ... : vector<4x512xf16>  // <[4, 64, 8], [BATCH, LANE, VECTOR]>
...
%6 = ... : vector<1x4x512xf16>  // <[2, 2, 64, 8], [BATCH, VECTOR, LANE, VECTOR]>
```
or
```
%6 = ... : vector<1x4x512xf16>  // <[2, 2, 64, 8], [0, 2, 3, 1], [BATCH, LANE, VECTOR, VECTOR]>
```
(or we don't worry about the multi-element reduction thing and let that be an implementation detail of lowering reduction, @kuhar let me know if this is something we have to represent as a part of the layout attribute).

For MFMA/tensor core stuff, this could simplify some of the instruction detection logic [here](https://github.com/openxla/iree/blob/dc4e59ec0c513872e320836142e79707edf28b1d/compiler/src/iree/compiler/Codegen/LLVMGPU/Utils/AMDGPUDistributionPatterns.cpp#L226). Instead of having to reconstruct an expected layout and compare, we instead just have to check that the inner most labels of the layout match the expected labels for the instruction.

As a final aside, as with the other layout, this layout is representable with an affine map, but the divs and mods required
to make it work are quite difficult to manage so my understanding is that there is an agreement to stay away from maps
as much as possible. Maybe in the end it turns out we need them still and then templating/interfaces can save the world,
but for now whatever we land on I'm +1 to trying to keep it simple.

**Comment by Groverkss** [comment]
@qedawkins How would you represent transpose and projection on your layout?

**Comment by qedawkins** [comment]
Transpose just transposes the entries in the layout the same way as before (i.e. apply the permutation to the expanded groups). Projection you just drop the "reduced" dims from the entries (by expanded bucket). It's the same as what's there now, just that instead of having the permutation of the data separated into buckets per dimension, it just lists the whole thing in one go. For me, I find this an easier layout to read than having to piece it together from each dimension.

**Comment by Groverkss** [comment]
Sure, I agree this is easier to read. I was just wondering how you are getting the `original dim --> expanded dims` mapping. It should be possible, but curious what you have in mind.

**Comment by qedawkins** [comment]
I see 3 ways:
1. Add a reassociation dims
2. Because vectors are statically shaped, add a helper to compute the reassociation dims from the expanded shape (just accumulate a product and such, maybe unit dims require the reassociation dims, but that means the layout is introducing unit dims for some reason).
3. Add additional reassociations to each of the arrays. Makes the permutation unwieldy if we do that, but could be a tradeoff.

Basically it's the same as what's there, but instead of `array<struct_with_4_entries>` it's `struct_with_4_arrays`.

I'm not overly attached to the exact representation above, this is mainly a suggestion after finding nested linear scans of nested attributes to find the `LANEX` dimension awkward to write code for. I see why it's nice for the analysis though, so there's probably a tradeoff.

**Comment by MaheshRavishankar** [comment]
So this looks very similar to the attributes we are using for encodings of packing. For packing, the encoding is really carry information on its "use", i.e. "is this the LHS of a matrix multiply operation", so its capturing the indexing maps and shapes of the users.
Not saying we should use that but was just giving an example. Are we interested in capturing "the use" of operation or not. (Just forming a mental model).

**Comment by Groverkss** [comment]
> So this looks very similar to the attributes we are using for encodings of packing. For packing, the encoding is really carry information on its "use", i.e. "is this the LHS of a matrix multiply operation", so its capturing the indexing maps and shapes of the users. Not saying we should use that but was just giving an example. Are we interested in capturing "the use" of operation or not. (Just forming a mental model).

The mental model is that the layout is an attribute on the type. So the layout is really attached to a SSA value. It is type information, which we carry through an analysis and some conflict ops because we cannot add an encoding on VectorType.

Different uses of a value have the same layout (since same type). If two uses have different layout expectation, there is an explicit conflict resolution operation which gives a value with a new layout.

```
a: #layout

use1 a  --> expects #layout
use2 a  --> expects #layout
```

```
a: #layout

use1 a --> expects #layout

a_x = layout_conflict a

use2 a_x --> expects #layout2
```

**Comment by qedawkins** [comment]
> So this looks very similar to the attributes we are using for encodings of packing. For packing, the encoding is really carry information on its "use", i.e. "is this the LHS of a matrix multiply operation", so its capturing the indexing maps and shapes of the users. Not saying we should use that but was just giving an example. Are we interested in capturing "the use" of operation or not. (Just forming a mental model).

We could, but I don't think the "use" matters outside of cooperative matrix (we probably do have to do it there, but that will require more thought). But yeah, basic idea is that it's similar to the packing layout (which can be decomposed into a sequence of tensor ops, this can be decomposed into a sequence of vector ops, sorta).

Edit: actually maybe it could matter, but I'm not sure the layout analysis is built to use it (Kunwar can correct me).

**Comment by harsh-nod** [comment]
Just fyi, this document clearly outlines the WMMA layout which we should be able to represent with our layout. https://gpuopen.com/learn/wmma_on_rdna3/

**Comment by Groverkss** [comment]
Closing this since we ended with NestedLayoutAttr, which is working quite well for us right now
